File: rs_impl/space_simulation/src/Planet.rs
use super::VectorMath as vm;

const GRAV_CONST: f32 = 0.0001; 

pub struct Planet {
    pub v: [f32; 2],
    pub pos: [f32; 2],
    pub name: String,
    pub color: (u8, u8, u8),
    pub radius: i16,
    pub surface_grav: f32,
}

impl Planet {
    pub fn get_radius(&self) -> i16 {
        return self.radius;
    }

    pub fn get_pos(&self) -> Vec<f32> {
        Vec::from(self.pos)
    }

    pub fn set_pos(&mut self, x: f32, y: f32) {
        self.pos = [x, y];
    }

    pub fn get_color(&self) -> (u8, u8, u8) {
        self.color
    }

    pub fn set_color(&mut self, color: (u8, u8, u8)) {
        self.color = color;
    }

    pub fn get_mass(&self) -> f32 {
        self.surface_grav * self.radius as f32 * self.radius as f32 / GRAV_CONST
    }

    pub fn get_vel(&self) -> Vec<f32> {
        let temp = Vec::from(self.v);
        vm::vec_nrm(temp)
    }

    pub fn update_velocity(&mut self, all_planets: Vec<Planet>, time_step: f32) {
        for planet in all_planets.iter() {
            let dst_sqr: f32 = vm::vec_dst(planet.get_pos(), self.get_pos());
            let force_dir: Vec<f32> = vm::vec_nrm(match vm::vec_sub(planet.get_pos(), self.get_pos()) {
                Some(v) => Vec::from(v),
                None => Vec::from([0., 0.])
            });
            // F = Gmm/r^2
            let force: Vec<f32> = vm::vec_mul(force_dir, GRAV_CONST * self.get_mass() * planet.get_mass() / dst_sqr);
            let acceleration: Vec<f32> = vm::vec_mul(force, self.get_mass());
            let temp: Vec<f32> = vm::vec_mul(acceleration, time_step);
            self.v[0] += temp[0];
            self.v[1] += temp[1];
        }
    }
}

File: py_impl/Planet.py
#! /usr/bin/env python3
import Universe

import numpy as np
import pygame

#--------------------------#
# Vector, scalar functions #
#--------------------------#

def vec_add(vec1, vec2):
    if (len(vec1) == len(vec2)):
        v = np.array([], dtype=float)
        for i in range(len(vec1)):
            v = np.append(v, vec1[i] + vec2[i])
        return v

def vec_sub(vec1, vec2):
    if (len(vec1) == len(vec2)):
        v = np.array([], dtype=float)
        for i in range(len(vec1)):
            v = np.append(v, vec1[i] - vec2[i])
        return v

def vec_sqt(x):
    v = np.array([], dtype=float)
    for i in range(len(x)):
        v = np.append(v, np.sqrt(x[i]))
    return v

def vec_mul(vec, scalar):
    v = np.array([], dtype=float)
    for i in vec:
        v = np.append(v, i * scalar)
    return v

def vec_div(vec, scalar):
    v = np.array([], dtype=float)
    for i in range(len(vec)):
        v = np.append(v, vec[i] / scalar)
    return v

def vec_pow(vec, pow):
    v = np.array([], dtype=float)
    for i in range(len(vec)):
        v = np.append(v, vec[i] ** pow)
    return v

def vec_mag(vec):
    sum: float = 0
    for i in vec:
        sum += i ** 2
    return np.sqrt(sum)

def vec_dst(vec1, vec2):
    if (len(vec1) == len(vec2)):
        sum: float = 0
        for i in range(len(vec1)):
            sum += (vec1[i] - vec2[i]) ** 2
        return sum
    return -1

def vec_nrm(vec):
    return vec_div(vec, vec_mag(vec))


#--------------#
# Planet Class #
#--------------#

class Planet():
    def __init__(self, name: str, pos: [float], vel: [float], radius: int, color: [int], grav: float):
        self.pos: [float] = np.array(pos, dtype=float)          # x and y position
        self.v: [float] = np.array(vel, dtype=np.float)     # x and y velocity
        self.name: str = name
        self.color: [int] = np.array(color, dtype=int)
        self.radius: int = radius
        self.surface_gravity:float = grav

    def get_radius(self):
        return self.radius

    def set_pos(self, x, y):
        self.pos[0] = x
        self.pos[1] = y
        return

    def get_pos(self):
        x = self.pos[0]
        y = self.pos[1]
        return (x, y)

    def set_color(self, r: int, g: int, b: int):
        self.color[0] = r
        self.color[1] = g
        self.color[2] = b

    def get_color(self):
        return self.color[0], self.color[1], self.color[2]

    def get_mass(self):
        return self.surface_gravity * self.radius * self.radius / Universe.GRAV_CONST

    def get_vel(self):
        return vec_nrm(self.v)

    def update_velocity(self, all_planets, time_step: float):
        for planet in all_planets:
            if planet != self:
                dst_sqr: float = vec_dst(planet.pos, self.pos)
                force_dir: [float] = vec_nrm(vec_sub(planet.get_pos(), self.get_pos()))
                force: [float] = force_dir * Universe.GRAV_CONST * self.get_mass() * planet.get_mass() / dst_sqr
                acceleration: [float] = force / self.get_mass()
                self.v += vec_mul(acceleration, time_step)

    def update_position(self, time_step: float):
        self.pos = vec_add(self.pos, vec_mul(self.v, time_step))
